go/arrow/compute/internal/kernels/cast_temporal.go

```diff
@@ -125,6 +129,10 @@ func TimestampToDate64(ctx *exec.KernelCtx, batch *exec.ExecSpan, out *exec.Exec
 
 	return ScalarUnaryNotNull(func(_ *exec.KernelCtx, arg0 arrow.Timestamp, _ *error) arrow.Date64 {
 		tm := fnToTime(arg0)
+		if _, offset := tm.Zone(); offset != 0 {
+			// normalize the tm
+			tm = tm.Add(time.Duration(offset) * time.Second).UTC()
+		}
```

Comment:
~this is effectively taking a timestamp with timezone and turning it into a naive timestamp showing the date in the former timezone (but marked as UTC)?

Reply:
Yes. Because as you pointed out above, when we call `.Unix()` it returns the result *not according to the location* but we want it to represent the date in the timezone's location, without having a timezone mark associated. This matches the C++ implementation of the casting in acero  (which is where I got the idea)

Reply:
Ok, cool. This just confuses me since I'm used to Java etc. where these are separate types ^_^